### PR TITLE
OTT-1659: Use NBR for logging APRC in logger

### PR DESCRIPTION
**Pull request by Pubmatic-Supriya-Patil**
# Description

Please add change description or link to ticket, docs, etc.

# Checklist:

- [ ] PR commit list is unique (rebase/pull with the origin branch to keep master clean).
- [ ] JIRA number is added in the PR title and the commit message.
- [ ] Updated the `header-bidding` repo with appropiate commit id.
- [ ] Documented the new changes.

For Prebid upgrade, refer: https://inside.pubmatic.com:8443/confluence/display/Products/Prebid-server+upgrade

### CODE REVIEW

File: endpoints/openrtb2/ctv/adpod/dynamic_adpod.go [hunk 1]
```diff
@@ -284,13 +286,13 @@ it will try to get the actual ad duration returned by the bidder using prebid.vi
 if prebid.video.duration not present then uses defaultDuration passed as an argument
 if video lengths matching policy is present for request then it will validate and update duration based on policy
 */
-func getBidDuration(bid *openrtb2.Bid, reqExt *openrtb_ext.ExtRequestAdPod, config []*types.ImpAdPodConfig, defaultDuration int64) (int64, constant.BidStatus) {
+func getBidDuration(bid *openrtb2.Bid, reqExt *openrtb_ext.ExtRequestAdPod, config []*types.ImpAdPodConfig, defaultDuration int64) (int64, *openrtb3.NoBidReason) {
```
Comment (ShriprasadM) [inline endpoints/openrtb2/ctv/adpod/dynamic_adpod.go:289]: @Pubmatic-Supriya-Patil  and @pm-nikhil-vaidya 
Missing here the SRP (Single Responsibility Principle). Its doing following 3 things

1.  Determining duration of the bid (which is its core responsibility)
2. Validating and determining the duration as per policy (not a responsibility of this function)
3. Determining Non-bid reason code. 

Refactored code 
```go
// validate bid duration
	bidDuration := getDuration(bid, da.ImpConfigs[sequence-1].MaxDuration)
	var durations []int
	for _, config := range da.ImpConfigs {
		durations = append(durations, int(config.MinDuration))
		durations = append(durations, int(config.MaxDuration))
	}
	matches, bidDuration := isDurationMatchesPolicy(bidDuration, da.ReqAdpodExt.VideoAdDurationMatching, durations)
	var nbr openrtb3.NoBidReason
	if !matches {
		nbr = exchange.ResponseRejectedInvalidCreative
	}
```

`ad_duration.go`
```go
package adpod

import (
	"sort"

	"github.com/buger/jsonparser"
	"github.com/prebid/openrtb/v20/openrtb2"
	"github.com/prebid/prebid-server/v2/openrtb_ext"
)

func getDuration(bid *openrtb2.Bid, defaultDuration int64) int64 {
	if bid.Dur > 0 {
		return bid.Dur
	}
	if bid.Ext != nil {
		if duration, err := jsonparser.GetInt(bid.Ext, "prebid", "video", "duration"); err == nil && duration > 0 {
			return duration
		}
	}
	return defaultDuration
}

func isDurationMatchesPolicy(bidDuration int64, policy openrtb_ext.OWVideoAdDurationMatchingPolicy, durations []int) (bool, int64) {
	sort.Ints(durations)
	for _, duration := range durations {
		switch policy {
		case openrtb_ext.OWExactVideoAdDurationMatching:
			if bidDuration == int64(duration) {
				return true, bidDuration
			}
		case openrtb_ext.OWRoundupVideoAdDurationMatching:
			if int64(duration) >= bidDuration {
				return true, int64(duration)
			}
		}
	}
	return false, 0
}

```

**Review by ShriprasadM — CHANGES_REQUESTED** [review]

File: endpoints/openrtb2/ctv/response/adpod_generator.go [hunk 2]
```diff
@@ -362,19 +366,19 @@ func evaluate(bids [][]*types.Bid, indices [][]int, totalBids int, maxCategorySc
...
+					return nil, inext, jnext, ptrutil.ToPtr(exchange.ResponseRejectedCreativeCategoryExclusions)
 				}
 			}
 
 			//Domain
 			for _, domain := range bid.ADomain {
 				hbc.domainScore[domain]++
 				if hbc.domainScore[domain] > 1 && (hbc.domainScore[domain]*100/totalBids) > maxDomainScore {
-					return nil, inext, jnext, constant.StatusDomainExclusion
+					return nil, inext, jnext, ptrutil.ToPtr(exchange.ResponseRejectedCreativeAdvertiserExclusions)
```
Comment (ShriprasadM) [inline endpoints/openrtb2/ctv/response/adpod_generator.go:377]: @Pubmatic-Supriya-Patil : can't we use vanilla Prebid's method rather for pointer conversion - https://github.com/prebid/prebid-server/blob/99431ad83fa7546d8e8d60acc3feae77c815f37f/exchange/non_bid_reason.go#L19

**Review by ShriprasadM — CHANGES_REQUESTED** [review]

File: endpoints/openrtb2/ctv/adpod/dynamic_adpod.go (hunk 1 above)
Comment (Pubmatic-Supriya-Patil) [reply endpoints/openrtb2/ctv/adpod/dynamic_adpod.go:289]: @ShriprasadM As discussed we can have this changes in modules code as we are moving to modules.

File: endpoints/openrtb2/ctv/adpod/adpod.go [hunk 3]
```diff
@@ -79,12 +75,9 @@ func GetNonBidParamsFromTypesBid(bid *types.Bid, seat string) openrtb_ext.NonBid
 func addSeatNonBids(bids []*types.Bid) openrtb_ext.NonBidCollection {
```
[the remark this answers is not in the document]
Comment (Pubmatic-Supriya-Patil) [reply endpoints/openrtb2/ctv/adpod/adpod.go]: Done

**Review by ShriprasadM — APPROVED** [review]
LGTM

File: endpoints/openrtb2/ctv/adpod/structured_adpod_test.go [hunk 4]
```diff
@@ -5,8 +5,8 @@ import (
 	"testing"
```
Comment (pm-nikhil-vaidya) [inline endpoints/openrtb2/ctv/adpod/structured_adpod_test.go:5]: Can you please add test case where provide exclusion configs and call hold auction and check if NBR are properly getting updated

Comment (Pubmatic-Supriya-Patil) [reply endpoints/openrtb2/ctv/adpod/structured_adpod_test.go:5]: Done

**Pull request merged**